### PR TITLE
Python: modernize build files and metadata

### PR DESCRIPTION
**Pull request by llohse**
- add meson buildfiles
- add PEP517 and PEP621 definitions

Allows to install python bindings via pip install .

### CODE REVIEW

**Comment by tfarago** [comment]
Oh wow, thanks for that! Will test this next week.

**Comment by tfarago** [comment]
Unfortunately it doesn't work for me on Ubuntu 20.04 with meson-python 0.7 and python3.8.
```bash
Processing /home/tomas/dev/ufo-core/python
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [42 lines of output]
      The Meson build system
      Version: 0.53.2
      Source dir: /home/tomas/dev/ufo-core/python
      Build dir: /home/tomas/dev/ufo-core/python/.mesonpy-icm7pzpr/build
      Build type: native build
      Project name: ufo-python
      Project version: 1.1.0
      C compiler for the host machine: cc (gcc 9.4.0 "cc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0")
      C linker for the host machine: cc ld.bfd 2.34
      Host machine cpu family: x86_64
      Host machine cpu: x86_64
      Program python3 found: YES (/home/tomas/virtenvs/concert/bin/python)
      Found pkg-config: /usr/bin/pkg-config (0.29.1)
      Dependency python found: YES (pkgconfig)
      Run-time dependency pygobject-3.0 found: YES 3.36.0
      Run-time dependency ufo found: YES 0.16.0
      Build targets in project: 1
      
      Found ninja-1.10.2.git.kitware.jobserver-1 at /tmp/pip-build-env-kt57az30/overlay/bin/ninja
      + meson setup --native-file=/home/tomas/dev/ufo-core/python/.mesonpy-native-file.ini -Ddebug=false -Doptimization=2 --prefix=/usr /home/tomas/dev/ufo-core/python /home/tomas/dev/ufo-core/python/.mesonpy-icm7pzpr/build
      Traceback (most recent call last):
        File "/home/tomas/virtenvs/concert/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
          main()
        File "/home/tomas/virtenvs/concert/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/home/tomas/virtenvs/concert/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 130, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/tmp/pip-build-env-kt57az30/overlay/lib/python3.8/site-packages/mesonpy/__init__.py", line 910, in get_requires_for_build_wheel
          if not project.is_pure and platform.system() == 'Linux':
        File "/tmp/pip-build-env-kt57az30/overlay/lib/python3.8/site-packages/mesonpy/__init__.py", line 642, in is_pure
          for data_type, files in self._install_plan.items():
        File "/tmp/pip-build-env-kt57az30/overlay/lib/python3.8/site-packages/mesonpy/__init__.py", line 561, in _install_plan
          return self._info('intro-install_plan').copy()
        File "/tmp/pip-build-env-kt57az30/overlay/lib/python3.8/site-packages/mesonpy/__init__.py", line 555, in _info
          json.loads(file.read_text())
        File "/usr/lib/python3.8/pathlib.py", line 1236, in read_text
          with self.open(mode='r', encoding=encoding, errors=errors) as f:
        File "/usr/lib/python3.8/pathlib.py", line 1222, in open
          return io.open(self, mode, buffering, encoding, errors, newline,
        File "/usr/lib/python3.8/pathlib.py", line 1078, in _opener
          return self._accessor.open(self, flags, mode)
      FileNotFoundError: [Errno 2] No such file or directory: '/home/tomas/dev/ufo-core/python/.mesonpy-icm7pzpr/build/meson-info/intro-install_plan.json'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```
Do I need some specific versions of the packages like meson-python?

Ironically `pip install .` already works on master

**Comment by llohse** [comment]
Thanks for testing! The only reason I could think of is that your meson might be too old - you run 0.53. Could you try updating it?

Indeed, `pip install .` already works with the legacy setuptools build instructions. But pep517 and meson have a few advantages.
It is cleaner, one can configure the compiler through meson, and does not require to manually call pkg-config. But more importantly, it scales.

My main motivation is to allow installing ufo (including the filters) as a python package. It is relatively easy to accomplish with meson. That way, one would only have to clone the repository, call `pip install .`, and everything is installed relative to the python site-packages.

**Comment by tfarago** [comment]
Oh, I ran it in virtualenv where I have meson 0.63 and I had to uninstall the system wide version. Even then, `pip install -v .` includes this line:
`meson setup --native-file=/home/tomas/dev/ufo-core/python/.mesonpy-native-file.ini -Ddebug=false -Doptimization=2 --prefix=/usr /home/tomas/dev/ufo-core/python /home/tomas/dev/ufo-core/python/.mesonpy-pyokeex3/build`, I need to change `--prefix` to my virtual env, do you by chance know how to do it? It seems the docs are not "mature" enough because what I found didn't work (pip insatll --build-option/--global-option).

Also, I don't think it's the best idea to have the complete ufo installed under python's site packages. It's not a python library so I would not force installing it via python. Though I get the "oneliner" is very intriguing. We should discuss this as the merger progresses which I think will be after summer.

**Comment by llohse** [comment]
> Oh, I ran it in virtualenv where I have meson 0.63 and I had to uninstall the system wide version. Even then, `pip install -v .` includes this line:
> `meson setup --native-file=/home/tomas/dev/ufo-core/python/.mesonpy-native-file.ini -Ddebug=false -Doptimization=2 --prefix=/usr /home/tomas/dev/ufo-core/python /home/tomas/dev/ufo-core/python/.mesonpy-pyokeex3/build`, I need to change `--prefix` to my virtual env, do you by chance know how to do it? It seems the docs are not "mature" enough because what I found didn't work (pip insatll --build-option/--global-option).

That looks correct. You don't have to specify any options. It should automatically get installed into the activated venv.

For me it looks something like:
```
    + meson setup --native-file=/tmp/pip-req-build-x6solf2r/.mesonpy-native-file.ini -Ddebug=false -Doptimization=2 --prefix=/usr /tmp/pip-req-build-x6solf2r /tmp/pip-req-build-x6solf2r/.mesonpy-pwlconlc/build
    + meson compile
    + meson install --destdir /tmp/pip-req-build-x6solf2r/.mesonpy-pwlconlc/install
```
From the temporary installation directory, a wheel is build, which is then installed as usual.

I agree that the documentation of meson-python is lacking. Nevertheless, scipy is already using it in production and numpy is currently migrating to it. 

> Also, I don't think it's the best idea to have the complete ufo installed under python's site packages. It's not a python library so I would not force installing it via python. Though I get the "oneliner" is very intriguing. We should discuss this as the merger progresses which I think will be after summer.

I would not suggest to *force* installing it this way, but having the option is nice. If done properly, it should be completely transparent. xraylib (a C-library), for example, also provides the option. Also I wuold not attempt it prior to the merger. After, I might just try it locally and you can decide if you want to merge it upstream or not - which I would understand, too.

Regardless, feel free not not merge this PR. It took me 30 minutes and was a fun exercise. I, too, am still learning meson and meson-python. I just thought you might find it useful, given that the python ecosystem is moving towards pep517.

**Comment by tfarago** [comment]
No no, it's not that I don't want to merge it, I just want to make sure it works on our systems and I can easily install it. And also as you say, if we can keep the traditional build and have `pip install .` as an option they why not, of course. I want to keep this PR but we'll wait with the merge a bit.

Right now I "almost" got it working. The problem is that I get `_ufo.so` suffixed: `_ufo.cpython-38-x86_64-linux-gnu.so` and than of course I get `ModuleNotFoundError: No module named '_ufo'`. Does simple `import ufo` work for you?

**Comment by llohse** [comment]
try now

**Comment by tfarago** [comment]
Works, thanks! Still, do you know how to configure meson variables, which you'd normally do with "meson configure"? Now pip creates a temp directory which gets deleted afterwards, which is clean, but I don't know how to set up my project, e.g. the mentioned `--prefix` in case I want to install it to some other directory. Maybe they are clever and `pip install --prefix .` would work but general options don't. I had a quick look and it seems this feature is [very new](https://github.com/pypa/pip/issues/5771). Maybe there is a workaround you know about?

**Comment by llohse** [comment]
I'm happy to hear that!
It looks like configuring meson variables via meson-python is not yet implemented (see https://github.com/FFY00/meson-python/issues/108).

However, I don't think I understand your use case here. What path would you like to change? 
The installation prefix is set by pip and not by the wheel, which is generated by meson-python. You can also build a wheel using pypa build as follows:
```
$ python -m build .
$ unzip -l ufo-1.1.0-py3-none-linux_x86_64.whl
Archive:  ufo-1.1.0-py3-none-linux_x86_64.whl
  Length      Date    Time    Name
---------  ---------- -----   ----
      185  07-22-2022 19:44   ufo-1.1.0.dist-info/METADATA
       86  07-22-2022 19:44   ufo-1.1.0.dist-info/WHEEL
    18088  07-22-2022 17:44   ufo/_ufo.cpython-39-x86_64-linux-gnu.so
     4457  07-22-2022 17:44   ufo/__init__.py
       68  07-22-2022 17:44   ufo/numpy.py
     4364  07-22-2022 17:44   ufo/tomopy.py
      499  07-22-2022 17:44   ufo-1.1.0.dist-info/RECORD
---------                     -------
    27747                     7 files

```

**Comment by tfarago** [comment]
No I meant in general having some meson options, because especially when building C extensions there might be some necessary (not in our case now, just in general), with `--prefix` it's clear ;-)

**Comment by llohse** [comment]
I agree. That's why I opened the upstream issue ;-).

**Comment by matze** [comment]
Seems the upstream issue has been resolved, so you could go forward. By the way, I'm a huge proponent of proper build system backends rather than bespoke `setup.py` scripts, so why not drop it with this PR?

**Comment by llohse** [comment]
> Seems the upstream issue has been resolved, so you could go forward. By the way, I'm a huge proponent of proper build system backends rather than bespoke `setup.py` scripts, so why not drop it with this PR?

The `setup.py` would become redundant with this PR, but could still be used for legacy systems. Keep it or drop it.

**Comment by tfarago** [comment]
I want to re-visit this with @gabs1234 a bit later.

**Comment by tfarago** [comment]
> Seems the upstream issue has been resolved, so you could go forward. By the way, I'm a huge proponent of proper build system backends rather than bespoke `setup.py` scripts, so why not drop it with this PR?

Yes of course. `setup.py` will go.

**Comment by tfarago** [comment]
Before merging, can someone please shed some light on this? I have numpy installed globally and wanted to do this in a venv, so I tried to tell pip not to install dependencies and after `pip install -v --no-deps .` I see this:
```bash
Using pip 22.3.1 from /home/tomas/virtenvs/concert/lib/python3.8/site-packages/pip (python 3.8)
Processing /home/tomas/dev/ufo-core/python
  Installing build dependencies: started
  Running command pip subprocess to install build dependencies
  Collecting meson
    Using cached meson-0.64.1-py3-none-any.whl (895 kB)
  Collecting ninja
    Using cached ninja-1.11.1-py2.py3-none-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (145 kB)
  Collecting meson-python
    Using cached meson_python-0.11.0-py3-none-any.whl (63 kB)
  Collecting numpy
    Using cached numpy-1.23.5-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.1 MB)
  Collecting tomli>=1.0.0
    Using cached tomli-2.0.1-py3-none-any.whl (12 kB)
  Collecting pyproject-metadata>=0.5.0
    Using cached pyproject_metadata-0.6.1-py3-none-any.whl (7.4 kB)
  Collecting packaging>=19.0
    Using cached packaging-21.3-py3-none-any.whl (40 kB)
  Collecting pyparsing!=3.0.5,>=2.0.2
    Using cached pyparsing-3.0.9-py3-none-any.whl (98 kB)
  Installing collected packages: ninja, tomli, pyparsing, numpy, meson, packaging, pyproject-metadata, meson-python
  ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
  black 21.12b0 requires tomli<2.0.0,>=0.2.6, but you have tomli 2.0.1 which is incompatible.
  Successfully installed meson-0.64.1 meson-python-0.11.0 ninja-1.11.1 numpy-1.23.5 packaging-21.3 pyparsing-3.0.9 pyproject-metadata-0.6.1 tomli-2.0.1
```
It just goes ahead and installs numpy anyways so of course it doesn't work in the end. When I install numpy in the venv all works (but this wasn't the case until recently, but that's another issue).
Can someone tell me what am I doing wrong?

**Comment by tfarago** [comment]
Thanks for clarification. Just for future me or anyone else needing wondering/needing this: I managed to disable the build dependencies installation with `pip install --no-build-isolation .`

File: python/meson.build [hunk 1]
```diff
@@ -0,0 +1,10 @@
+project('ufo-python', 'c',
+        default_options : ['c_std=c99', 'warning_level=3'],
+        version : '1.1.0')
```
Comment (tfarago) [inline python/meson.build:3]: @llohse is this a copy-paste from something? I would perhaps use the same version as the `ufo-core` version.

Comment (llohse) [reply python/meson.build:3]: I took it from here: https://github.com/ufo-kit/ufo-core/blob/master/python/setup.py#L9

But yes...

**Comment by tfarago** [comment]
@llohse thanks again!

**Pull request merged**